src/components/listing-components/login-dialog.tsx

```diff
@@ -80,86 +87,211 @@ const loginDialog: React.FC<LoginDialogProps> = ({ open, onClose, onLogin }) =>
-  const handleLogin = () => {
+  const handleLogin = async (event: React.ChangeEvent<HTMLFormElement>) => {
+    event.preventDefault();
     const isValidLogin = validateLogin(username, password);
-  
     if (isValidLogin) {
-      onLogin();
-      onClose();
+      const { accessToken: newAccessToken, refreshToken: newRefreshToken, error } = await fetchLogin({
+        selectedSite: siteList.find(siteItem => siteItem.id === site),
+        username: username,
+        password: password
+      });
+
+      if (error) {
+        setLoginError(error);
+      } else {
+        setAccessToken(newAccessToken);
+        setRefreshToken(newRefreshToken);
+        onLogin();
+        onClose(loginError, undefined);
+      }
     } else {
       setLoginError(strings.errorHandling.listingScreenLogin.loginFailed);
     }
   };
 
   /**
+   * Token refresh logic
+   */
+  const tokenRefresh = async () => {
+    const { accessToken: newAccessToken, refreshToken: newRefreshToken, error } = await handleTokenRefresh({
+      selectedSite: siteList.find(siteItem => siteItem.id === site),
+      refreshToken: refreshToken
+    });
+    if (error) {
+      errorContext.setError(error);
+    } else {
+      setAccessToken(newAccessToken);
+      setRefreshToken(newRefreshToken);
+    }
+  };
+  
+  // Var for token refresh timer
+  let expirationTime = 0;
+  /**
+   * Check if token is expiring
+   * @param token 
+   * @returns 
+   */
+  const isTokenExpiring = (token: any) => {
+    if (!token || !token.expires_in) {
+      return false;
+    }
+
+    if (expirationTime === 0) {
+      expirationTime = Math.floor(Date.now() / 1000) + token.expires_in;
+    }
+
+    const currentTime = Math.floor(Date.now() / 1000); // Current time in seconds
+    const countdown = expirationTime - currentTime;
+    const refreshThreshold = 4 * 60; // 4 minutes
+
+    if (countdown <= 0) {
+      expirationTime = 0;
+      return true;
+    }
+    if (countdown <= refreshThreshold) {
+      expirationTime = currentTime + token.expires_in;
+      return true;
+    }
+
+    return countdown <= refreshThreshold;
+  };
+
+  React.useEffect(() => {
+    /**
+     * Initialize IsTokenExpiring
+     */
+    const initializeTokenExpiration = () => {
+      if (isTokenExpiring(accessToken)) {
+        setTokenExpiring(true);
+        expirationTime = 0;
+      }
+    };
+    initializeTokenExpiration();
+  });
+  
+  /**
+   * Reset the timer on token expiration time
+   */
+  React.useEffect(() => {
+    const refreshTimeout = setTimeout(() => {
+      if (isTokenExpiring(accessToken)) {
+        setTokenExpiring(true);
+        expirationTime = 0;
+      }
+    }, 3 * 60 * 1000);
+
+    return () => {
+      clearTimeout(refreshTimeout);
+    };
+  }, [accessToken]);
+
+  /**
+   * Check if token is expiring --> refresh token || false
+   */
+  React.useEffect(() => {
+    if (tokenExpiring) {
+      tokenRefresh();
+      setTokenExpiring(false);
+    }
+  }, [tokenExpiring]);
```

Comment:
This is unnecessarily complex, you can make do with just one useEffect, that adds and cleans up an interval via setInterval() and clearInterval() with the correct functionality.

There is one catch, though. You will need to update the callback in the setInterval with the updated token after every refresh. This can be done easily e.g. with a handy React hook "useInterval". There are many implementations of it in the web in different libraries, but I can provide one I created myself here:

```
/**
 * Custom hook for running given callback function in intervals
 *
 * @param callback callback function
 * @param interval interval as milliseconds
 */
export const useInterval = (callback: () => any, interval: number) => {
  const savedCallback = useRef<typeof callback>();

  useEffect(() => {
    savedCallback.current = callback;
  });

  useEffect(() => {
    const tick = () => {
      savedCallback.current?.();
    };

    const timeout = setInterval(tick, delay);
    return () => clearInterval(timeout);
  }, [interval]);
};
```

This you can use it inside a React component just like this:
```
useInterval(() => {
  ...callback logic...
}, 3 * 60 * 1000);
```
Inside that you can just check whether the token will soon expire, and do the refresh.